---
title: "Analyses_October30"
author: "Ana Miller-ter Kuile"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 2
    df_print: paged
    theme: readable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r, include = FALSE}
package.list <- c("here", "tidyverse", 
                  "glmmTMB", "emmeans",
                  "MuMIn", "DHARMa",
                  "effects", "ggeffects",
                  "patchwork", "calecopal",
                  "gt", "vegan", "remotes",
                  "eulerr", "dummies")

## Installing them if they aren't already on the computer
new.packages <- package.list[!(package.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

## And loading them
for(i in package.list){library(i, character.only = T)}
```


# Inter- and intraspecific body size patterns in individual predator-prey interactions from DNA metabarcoding

The analyses for these projects have been inspired by ideas and approaches outlined in: [Brose et al. 2019](https://doi.org/10.1038/s41559-019-0899-x), [Rudolf et al. 2014](https://doi.org/10.1098/rspb.2013.3203), and [Woodward and Hildrew 2002](https://doi.org/10.1046/j.1365-2656.2002.00669.x), among others. Specifically, these analyses explore the importance of body size (both within and across species) and species identity in shaping food web patterns. 

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "geophilomorpha.png"))
```

## Introduction

Food webs are regularly built using "nodes" based on size, predator identity, or a combination of size and identity (i.e. "stage structure"). For food webs, nodes are created by a combination of predator species and size and the interactions (links) between predators and their prey items are determined using size-based rules related to gape limitation. Across species, sizes, and environments, there seem to be consistent patterns of size relationships in food webs - for example, variation in the relationships between predator and prey sizes can be explained by predator traits (e.g. locomotion, thermoregulation) across 290 food webs (Brose et al. 2019). While these patterns are promising for the efficient construction of food webs across many orders of magnitude of body size, they have still not been validated by real data for many species of smaller body sizes, including invertebrates in terrestrial environments. 

In several aquatic field and mesocosm studies, it has been clear that both the body size and species identity of predators is important in determining both the identity and body size of prey items (e.g. Woodward and Hildrew 2002, Rudolf et al. 2014). However, for terrestrial predators, which comprise (%) of total species on the planet, determining these size rules of predation based on observed interactions has been challenging or impossible. These consumers, thus, constitute a data gap of predator-prey interactions built on observed interaction data.

To aim to fill this gap in observed interactions for invertebrate terrestrial predators, in this study, we employed DNA metabarcoding of the COI gene region using general primers for all animals (following an adapted protocol from Krehenwinkel et al. 2016 and Miller-ter Kuile et al. *in review*) to determine the diets of nine species of invertebrate predators on Palmyra Atoll, Central Tropical Pacific. We determined interactions for 182 predator individuals, encompassing 3.5 orders of magnitude of body size and including 339 distinct interactions. Using individual-level body size data for each predator individual and averaged body size data for prey items (similar to resolution for prey in other food webs e.g. Brose et al. 2019), we asked three questions relating to body size in food webs. 1) Does predator size, species identity, or their combination determine prey size (level: interspecific)?, 2) Does predator size or identity determine prey identity for predator individuals from the same environment (level: interspecific)?, and 3) Within a predator species, do smaller individuals eat a subset of the prey of larger predator individuals (e.g. nestedness or a "nested hierarchy", Woodward and Hildrew 2002, level: intraspecific)? These three questions are key to building multi-species food web models that incorporate realistic body size and species dependent feeding habits both across and within predator species.

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "neoscona.png"))
```

## The data

This dataset represents interactions between individual predators and their prey. The predators range in size from $2.3 x 10^{-1}$ to $9.3 x 10^2$ (3 orders of magnitude). Predators represent 9 species, including five species of spider (*Heteropoda venatoria*, *Neoscona theisi*, *Ooonopidae sp.*, *Scytodes longipes*, *Smeringopus pallidus*), one centipede (*Geophilomoropha sp.*), one earwig (*Euborellia annulipes*), a predatory katydid (*Phisis holdhausi*), and a dragonfly species (*Pantala flavescens*).

```{r, include = FALSE}
source(here("6_statistics", "1_pred_prey_size.R"))
```

### Predator size distribution

```{r}
pred_size2
```

### Predator size distribution by species

```{r}
pred_labels <- c("CEN" = "Geophilomorpha sp.", "EUB" = "E. annulipes", 
                 "HEV" = "H. venatoria", "LRS" = "Oonopidae sp.", 
                 "NEO" = "N. theisi", "PAN" = "P. flavescens",
                 "PHH" = "P. holdhausi", "SCY" = "S. longipes",
                 "SME" = "S. pallidus")

size %>%
  distinct(sample, sample_str, pred_mass_mg) %>%
  ggplot(aes(x = pred_mass_mg, fill = sample_str)) +
  geom_histogram(color = "black", alpha = 0.85) +
  theme_bw() + 
  scale_x_log10() +
  labs(x = "Predator mass (mg)", y = "Individuals") +
    theme(axis.text = element_text(size =20),
          axis.title = element_text(size = 25)) +
  facet_wrap(~sample_str, labeller = labeller(sample_str = pred_labels)) +
  scale_fill_manual(values = pal_kelp)
```

I compiled prey DNA data at the family level because of resolution in online taxonomy databases. Summarising prey at the family level is common for food webs of interactions with terrestrial invertebrates (e.g. Brose et al. 2019). The prey family range in size from $6.3 x 10^{-4}$ to $3.1 x 10^2$ (6 orders of magnitude). 
 
### Mean prey size distribution

```{r}
prey_mean_size2
```

### Minimum prey size distribution

```{r}
prey_min_size2
```

Prey comprise 57 different families of invertebrate organisms. 

```{r}
size %>%
  distinct(Class, Order, Family) %>%
  arrange(Class, Order, Family)
```

## Q1: Predator-prey size relationships

### Does predator size, species identity, or their combination determine prey size (level: intraspecific)?

For this question, I used LMMs with log-transformed size values (due to data distribution) and performed model selection using AIC comparing models to a full model that included the interaction between predator body size and species identity (prey_size ~ predator_size x predator_species).

The outcome was that a model with predator_size + predator_species without the interaction was the best-fitting model. This would mean that larger individuals eat larger prey (with a log-log relationship in this case of $y = x^{0.41}$) and that the intercept of this model varies by species. However, I have a few questions moving forward with this analysis.

#### Unaddressed challenges with this approach

#### 1. Size and species are co-linear

```{r}
size %>%
  distinct(sample, sample_str, pred_mass_mg) %>%
  mutate(sample_str = factor(sample_str, 
                             levels = c("LRS","NEO", "CEN", "EUB",
                                        "SCY", "SME", "PHH", "PAN", "HEV"))) %>%
  ggplot(aes(x = sample_str, y = pred_mass_mg)) +
  geom_boxplot() +
  scale_y_log10() +
  labs(x = "Predator species", y = "Predator mass (mg)") +
  theme_bw()
```
Hence, a model structure with an interaction (prey_size ~ predator_size x predator_species) seems easy-ish to interpret (e.g. predator species changes the *slope* and *intercept* of the predator-prey body size relationship) but a model structure with just predator species and predator size seems... less easy to interpret. This seems to suggest an cross-species relationship between predator-prey body size, but says less about within species. Would love any thoughts you have on this!

#### 2. How to visualize results when raw data are... a  blob?

The interpretation of each model structure is as follows:

1. Interaction model (prey_size ~ predator_size x predator_species): Both predator size and species identity determine prey size and both the **slope** and **intercept** of this model vary by predator species.

2. Size + species model (prey_size ~ predator_size + predator_species): Both predator size and species identity determine prey size and only the **intercept** of this model varies by predator species.

3. Size model (prey_size ~ predator_size): There is a relationship between prey size and predator size and *predator species does not change this relationship*.

4. Species model (prey_size ~ predator_species): *Regardless of predator size*, each predator species eats a distinct size range of prey. 

I ran model selection both with the **mean size** and **minimum size** of each prey family as the response variable. In both cases, the best model based on AIC was the **Size + Species model**, which means that there is a significant (positive) relationship between prey size and predator size; the slope of this line does not change across predator species, but each species has a different intercept for this relationship. Not surprisingly, the slope of the relationship varied between the mean size and minimum size models, with the slope being shallower for minimum prey size. 

### Mean prey size

```{r}
mean_predprey_plot
```

### Minimum prey size

```{r}
min_predprey_plot
```

#### Predator-prey size interpretations

Both predator species and predator size is an important determinant of prey size, with an invariant slope of this relationship with varying intercepts by species. This means that when building food webs, both species and body size are important determinants of deciding on nodes and links at the species or species-stage level. For both mean and minimum models above, while there is some shifting in the predators with the middle intercepts, the predators with the lowest and highest intercepts remained the same. From what I know of these predators' traits (e.g. locomotion, hunting strategy) there really doesn't seem to be any clear trait-based reason for these differences. Also, given that I did not collect these data specifically targeting predators of a variety of traits while correcting for others, I really don't know what we could say here at the trait level. That said, knowing that species identity is an important co-variate when building size-based food webs is an important finding (e.g. Rudolf et al. 2014). I also think a cool "next step" study could be to look at these relationships for a set of predators with shared size and different hunting strategies or locomotion (e.g. inspiration from Schmitz 2009).

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "goblin.png"))
```

## Q2: Predator-prey identity and size

### Within a shared environment, does predator size or identity determine prey identity?

Next, I wanted to know whether for predators in a shared environment, prey item identity is determined by predator size or predator species identity. To do this, I subset just the predators I had collected from one shared environment and sampling period (a canopy fogging collection in *Pisonia grandis* on Sand Island in 2017). In this particular fogging event, I was able to collect 60 individuals from four predator species (*H. venatoria*, *P. holdhausi*, *N. theisi*, and *E. annulipes*).

```{r, include = FALSE}
source(here("6_statistics", "3_cca_sandpg_canopy.R"))
```

### Size distribution of predators from Sand *Pisonia grandis*

```{r}
cols <-  c("EUB" = "#EA7700", "HEV" = "#EEB00C","NEO" = "#89742F", "PHH" = "#158D8E")

size2 %>%
  filter(Island == "Sand" & 
           Habitat == "PG" & 
           Microhabitat == "Canopy") %>%
  distinct(sample, sample_str, pred_log_mass_mg) %>%
  ggplot(aes(x = pred_log_mass_mg, fill = sample_str)) +
  geom_histogram(color = "black", position = "identity", alpha = 0.8) +
  theme_bw() +
  scale_fill_manual(values = cols)
```


These predators have eaten 24 different prey families.

```{r}
size2 %>%
  filter(Island == "Sand" & 
           Habitat == "PG" & 
           Microhabitat == "Canopy") %>%
  distinct(Class, Order, Family) %>%
  arrange(Class, Order, Family) 
```

To determine whether these predators were partitioning prey by size, species, or both, I ran a constrained canonical analysis (CCA) with both predator species identity and size as predictor variables. CCA is a multivariate method that tests for relationships between predictor variables and a set of multivariate data (e.g. a set of diet "communities" across individual predator samples). Unlike RDA, CCA does not assume a linear relationship and is more aimed at understanding how variables influence community composition. 

I first performed a full model including both predator size and species identity. This model shows a significant influence of one or more model terms on prey species composition.

```{r}
#ANOVA of whole model
anova(cca_pg, permutations=10000)

#how much variation explained in this RDA
RsquareAdj(cca_pg)
```

When examining the terms of this model, both predator species and predator size are important predictors of community composition.

```{r}
#ANOVA of model terms
anova(cca_pg, by='margin', permutations=10000)
```

```{r}
cca_pg
```

From this output, we see that the variables of predator species and predator size explain 19% of the variation in the data. Furthermore, of this variation, CCA1 explains 83.3% of the variation and CCA2 explains 67.0% of this variation. 

### CCA and of prey composition by predator

```{r}
CCA_plot
```

We can go further and visualize the variation due to predator species and predator size using an Euler plot, which is like a Venn diagram in that it shows the contribution to variation in each variable independent and dependent on other variables. 

```{r}
euler
```

The Euler plot shows that predator species explains 16.2% of the variation alone, while predator size within species explains an additional 3.4%. Predator size alone contributes a negligible amount of explanatory power (0.1%) to the relationship. 

#### Predator-prey identity and size interpretation

Predator species and size both explain the distribution of samples ('sites') along the main axes of the CCA. Whether a predator is **EUB** (*Euborellia annulipes*) or **PHH** (*Phisis holdhausi*) and how large a predator is predicts much of the variation along the CCA1 axis (the majority of variation in the CCA), whereas whether a predator was **HEV** (*Heteropoda venatoria*) or **NEO** (*Neoscona theisi*) explains much of the variation on the CCA2 axis. The result of this CCA indicates that in a shared environment, predators seem to be partitioning resources by species with some smaller contribution based on predato size.

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "phisis.png"))
```

## Q3: Within species prey identities

### Within a species, do smaller individuals eat a subset of the prey of larger individuals (e.g. nestedness)?

So far, I've been examining individual-level patterns across species of predators. However, another aspect of size-based food webs includes *intraspecific* selection of prey items. In many consumers, there is an apparent nested hierarchy or nestedness to intraspecific diet in which smaller individuals eat a subset of the same prey items that larger individuals of that species consume. This is one of the theoretical bases of the niche model for building null food webs (CITE these). I wanted to see if I could detect any sort of nestedness in these data. I selected three predator species for which I have a larger sample size (24 - 53 individuals within the species). These predator species included *H.venatoria* (53 individuals), *N. theisi* (24 individuals), and *P. holdhausi* (42 individuals). 

To do this, I asked whether the diets of individuals were nested compared to random simulations of the same community (I used the NODF nestedness measure, Ulrich et al. 2009), which incorporates pairwise similarities across sites and species occurrences and is invariant to the size of the sample or community sampled. Specifically, the NODF nestedness value can take community assemblies and ask whether the most species poor of these comprise a subset of the species that are present in a richer community (Almeida-Neto et al. 2008). 

```{r}
source(here("6_statistics", "4_intraspecific_nestedness.R"))
```

### *H. venatoria* nestedness analysis

```{r}
oecosimu(mat_hev, nestednodf, "quasiswap")
```

The results of this simulation indicate that the prey communities across *H. venatoria* individuals are not significantly nested compared to randomly-constructed communities.

A visualization of this:

```{r}
hev_matrix 
```

And suggestion that we sampled an estimated 28 of an estimated 42 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
hev_acc
```

### *N. theisi* nestedness analysis

```{r}
oecosimu(mat_neo, nestednodf, "quasiswap")
```

Again, the results of this simulation indicate that the prey communities across *N. theisi* individuals are not significantly nested compared to randomly-constructed communities.

```{r}
neo_matrix
```

And suggestion that we sampled an estimated 29 of an estimated 44 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
neo_acc
```

### *P. holdhausi* nestedness analysis

```{r}
oecosimu(mat_phh, nestednodf, "quasiswap")
```

Again, the results of this simulation indicate that the prey communities across *P. holdhausi* individuals are not significantly nested compared to randomly-constructed communities.

```{r}
phh_matrix
```

And suggestion that we sampled an estimated 18 of an estimated 35 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
phh_acc
```

#### Predator-prey identity and size interpretation

None of these predator populations seem to have significantly nested diets within a species. Unlike results from e.g. Woodward and Hildrew 2002, it does not seem like a good assumption on building feeding interactions within species stages would be to assign a subset of prey to smaller individuals than larger individuals. We captured anywhere from 51 - 67% of total prey richness based on accumulation curves and I'm wondering if this is coming into play here. 

Any other ways to think about intraspecific size? I played around with some hierarchical clustering algorithms, which could be promising, but wondering if folks have other ideas from the literature on how to think about intraspecific diet!

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "pantala.png"))
```


## Comments from co-authors/myself:

1. How to visualize/interpret predator-prey body size data with and without a log-log scale?

2. **Austen**: Is there a reason you need to control for the environment in Q2 but not Q1? It's probably obvious, but I still can't think of why. Couldn't you run with the CCA in Q2 with environment-type (canopy vs. non-canopy) as a covariate? Then you could include everything? Relatedly, is there a reason you only ran this analysis with canopy predators as a subset and not understory predators as another analysis? 

3. **Austen**: Would it be interesting to use variation in prey mass as a response variable to complement results from Q1/Q3? I think this could complement your nestedness analysis for Q3 where you found no nestedness related to prey species identity. Nestedness could occur by prey identity irrespective of prey size (NODF analysis) but nestedness (in a different sense) could also occur by prey size irrespective of prey identity, right? The same analysis you ran for Q1 with prey mass SD as the response variable could tell you this perhaps. In other words, larger H. venatoria spiders may have a wider prey mass SD bc they like to eat potato chips (small prey) and steaks (large prey; though they tend towards steaks - results from Q1), while smaller H. venatoria spiders still eat potato chips (small prey), but eat a different brand of potato chips than larger H. venatoria, hence, you would not see an effect of nestedness by prey identity with a NODF analysis but could see an effect with an LMM. Let's hope that made sense... 

4. (not really sure what is being to referred to here, **Austen**: To your point about using abundance-weighted data, I think it would make good sense to try it with abundance-weighted data.)

5. Which predator codes are for Ooonopidae sp. and Geophilomoropha sp.? 


