---
title: "Analyses_October30"
author: "Ana Miller-ter Kuile"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 2
    df_print: paged
    theme: readable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r, include = FALSE}
package.list <- c("here", "tidyverse", 
                  "glmmTMB", "emmeans",
                  "MuMIn", "DHARMa",
                  "effects", "ggeffects",
                  "patchwork", "calecopal",
                  "gt", "vegan", "remotes",
                  "eulerr", "dummies")

## Installing them if they aren't already on the computer
new.packages <- package.list[!(package.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

## And loading them
for(i in package.list){library(i, character.only = T)}
```


# Inter- and intraspecific body size patterns in individual predator-prey interactions from DNA metabarcoding

The analyses for these projects have been inspired by ideas and approaches outlined in: [Brose et al. 2019](https://doi.org/10.1038/s41559-019-0899-x), [Rudolf et al. 2014](https://doi.org/10.1098/rspb.2013.3203), and [Woodward and Hildrew 2002](https://doi.org/10.1046/j.1365-2656.2002.00669.x), among others. Specifically, these analyses explore the importance of body size (both within and across species) and species identity in shaping food web patterns. 

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "geophilomorpha.png"))
```

## Introduction

Food webs are regularly built using "nodes" based on size, predator identity, or a combination of size and identity (i.e. "stage structure"). For food webs, nodes are created by a combination of predator species and size and the interactions (links) between predators and their prey items are determined using size-based rules related to gape limitation. Across species, sizes, and environments, there seem to be consistent patterns of size relationships in food webs - for example, variation in the relationships between predator and prey sizes can be explained by predator traits (e.g. locomotion, thermoregulation) across 290 food webs (Brose et al. 2019). While these patterns are promising for the efficient construction of food webs across many orders of magnitude of body size, they have still not been validated by real data for many species of smaller body sizes, including invertebrates in terrestrial environments. 

In several aquatic field and mesocosm studies, it has been clear that both the body size and species identity of predators is important in determining both the identity and body size of prey items (e.g. Woodward and Hildrew 2002, Rudolf et al. 2014). However, for terrestrial predators, which comprise (%) of total species on the planet, determining these size rules of predation based on observed interactions has been challenging or impossible. These consumers, thus, constitute a data gap of predator-prey interactions built on observed interaction data.

To aim to fill this gap in observed interactions for invertebrate terrestrial predators, in this study, we employed DNA metabarcoding of the COI gene region using general primers for all animals (following an adapted protocol from Krehenwinkel et al. 2016 and Miller-ter Kuile et al. *in review*) to determine the diets of nine species of invertebrate predators on Palmyra Atoll, Central Tropical Pacific. We determined interactions for 182 predator individuals, encompassing 3.5 orders of magnitude of body size and including 339 distinct interactions. Using individual-level body size data for each predator individual and averaged body size data for prey items (similar to resolution for prey in other food webs e.g. Brose et al. 2019), we asked three questions relating to body size in food webs. 1) Does predator size, species identity, or their combination determine prey size (level: interspecific)?, 2) Does predator size or identity determine prey identity for predator individuals from the same environment (level: interspecific)?, and 3) Within a predator species, do smaller individuals eat a subset of the prey of larger predator individuals (e.g. nestedness or a "nested hierarchy", Woodward and Hildrew 2002, level: intraspecific)? These three questions are key to building multi-species food web models that incorporate realistic body size and species dependent feeding habits both across and within predator species.

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "neoscona.png"))
```

## The data

This dataset represents interactions between individual predators and their prey. The predators range in size from $2.3 x 10^{-1}$ to $9.3 x 10^2$ (3 orders of magnitude). Predators represent 9 species, including five species of spider (*Heteropoda venatoria*, *Neoscona theisi*, *Ooonopidae sp.*, *Scytodes longipes*, *Smeringopus pallidus*), one centipede (*Geophilomoropha sp.*), one earwig (*Euborellia annulipes*), a predatory katydid (*Phisis holdhausi*), and a dragonfly species (*Pantala flavescens*).

```{r, include = FALSE}
source(here("6_statistics", "1_pred_prey_size.R"))
```

### Predator size distribution

```{r}
pred_size2
```

### Predator size distribution by species

```{r}
pred_labels <- c("CEN" = "Geophilomorpha sp.", "EUB" = "E. annulipes", 
                 "HEV" = "H. venatoria", "LRS" = "Oonopidae sp.", 
                 "NEO" = "N. theisi", "PAN" = "P. flavescens",
                 "PHH" = "P. holdhausi", "SCY" = "S. longipes",
                 "SME" = "S. pallidus")

size %>%
  distinct(sample, sample_str, pred_mass_mg) %>%
  ggplot(aes(x = pred_mass_mg, fill = sample_str)) +
  geom_histogram(color = "black", alpha = 0.85) +
  theme_bw() + 
  scale_x_log10() +
  labs(x = "Predator mass (mg)", y = "Individuals") +
    theme(axis.text = element_text(size =20),
          axis.title = element_text(size = 25)) +
  facet_wrap(~sample_str, labeller = labeller(sample_str = pred_labels)) +
  scale_fill_manual(values = pal_kelp,,
                     labels = pred_labels) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

I compiled prey DNA data at the family level because of resolution in online taxonomy databases. Summarising prey at the family level is common for food webs of interactions with terrestrial invertebrates (e.g. Brose et al. 2019). The prey family range in size from $6.3 x 10^{-4}$ to $3.1 x 10^2$ (6 orders of magnitude). 
 
### Mean prey size distribution

```{r}
prey_mean_size2
```

### Minimum prey size distribution

```{r}
prey_min_size2
```

Prey comprise 57 different families of invertebrate organisms. 

```{r}
size %>%
  distinct(Class, Order, Family) %>%
  arrange(Class, Order, Family)
```

## Q1: Predator-prey size relationships

### Does predator size, species identity, or their combination determine prey size (level: interspecific)?

For this question, I used LMMs with log-transformed size values (due to data distribution) and performed model selection using AIC comparing models to a full model that included the interaction between predator body size and species identity (prey_size ~ predator_size x predator_species). I ran model selection both with the **mean size** and **minimum size** of each prey family as the response variable.

The outcome for both response variables was that a model with predator_size + predator_species without the interaction was the best-fitting model. This would mean that larger individuals eat larger prey (with a log-log relationship in this case of $y = x^{0.41}$ for *mean prey size* and $y = x^{0.20}$ for *min prey size*) and that the intercept of this model varies by species. However, I have a few questions moving forward with this analysis.

### Unaddressed challenges with this approach

#### 1. Size and species are co-linear

```{r}
size %>%
  distinct(sample, sample_str, pred_mass_mg) %>%
  mutate(sample_str = factor(sample_str, 
                             levels = c("LRS","NEO", "CEN", "EUB",
                                        "SCY", "SME", "PHH", "PAN", "HEV"))) %>%
  ggplot(aes(x = sample_str, y = pred_mass_mg)) +
  geom_boxplot() +
  scale_y_log10() +
  labs(x = "Predator species", y = "Predator mass (mg)") +
  scale_x_discrete(labels = pred_labels) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


Hence, a model structure with an interaction (prey_size ~ predator_size x predator_species) seems easy-ish to interpret (e.g. predator size *within* species changes the *slope* and *intercept* of the predator-prey body size relationship) but a model structure with just predator species and predator size seems... less easy to interpret, especially given this co-linearity. This seems to suggest a cross-species relationship between predator-prey body size, but says less about within species. **Would love any thoughts you have on this!**

#### 2. How to visualize raw data that are... a  blob?

The raw data (with dashed line indicating 1:1 size ratios):

#### Mean prey size

```{r}
ggplot(size, aes(x = pred_mass_mg, y = mean_prey_mass_mg, color = sample_str)) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 3) +
  #scale_x_log10() +
  #scale_y_log10() +
  theme_bw() +
  #facet_wrap(~sample_str, scale = "free", labeller = labeller(sample_str = pred_labels)) +
  scale_color_manual(values = pal_kelp,
                     labels = pred_labels)
```

#### Log-log mean prey size

```{r}
ggplot(size, aes(x = pred_mass_mg, y = mean_prey_mass_mg, color = sample_str)) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 3) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw() +
  #facet_wrap(~sample_str, scale = "free", labeller = labeller(sample_str = pred_labels)) +
  scale_color_manual(values = pal_kelp,
                     labels = pred_labels)
```


#### Min prey size

```{r}
ggplot(size, aes(x = pred_mass_mg, y = min_prey_mass_mg, color = sample_str)) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 3) +
  #scale_x_log10() +
  #scale_y_log10() +
  theme_bw() +
  #facet_wrap(~sample_str, scale = "free", labeller = labeller(sample_str = pred_labels)) +
  scale_color_manual(values = pal_kelp,
                     labels = pred_labels)
```

#### Log-log min prey size

```{r}
ggplot(size, aes(x = pred_mass_mg, y = min_prey_mass_mg, color = sample_str)) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 3) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw() +
  #facet_wrap(~sample_str, scale = "free", labeller = labeller(sample_str = pred_labels)) +
  scale_color_manual(values = pal_kelp,
                     labels = pred_labels)
```

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "goblin.png"))
```

## Other questions/approaches (not covered in detail here)

I need the most help with Q1 above, but if you have thoughts about the following questions or approaches, I would love to hear them! Thank you!

2. **Q2: Within a shared environment, does predator size or identity determine prey identity?**

  + Approach: CCA of predator prey identity by species and body size
  
  + Findings: predator identity determines prey identity with a small contribution of size within predator species.
  
  + Next steps: **Austen**: Is there a reason you need to control for the environment in Q2 but not Q1? It's probably obvious, but I still can't think of why. Couldn't you run with the CCA in Q2 with environment-type (canopy vs. non-canopy) as a covariate? Then you could include everything? Relatedly, is there a reason you only ran this analysis with canopy predators as a subset and not understory predators as another analysis? 

3. **Q3: Within a species, do smaller individuals eat a subset of the prey of larger individuals (e.g. nestedness)?**

  + Approach: nestedness analysis of prey identity by predator individuals within a species using NODF, which scores a group of sites (individuals) on whether the most common species are present in the least species rich sites and whether the rarest species are also in sites with the most species richness
  
  + Findings: no real evidence of nestedness of prey
  
  + Next steps: **Austen**: Would it be interesting to use variation in prey mass as a response variable to complement results from Q1/Q3? I think this could complement your nestedness analysis for Q3 where you found no nestedness related to prey species identity. Nestedness could occur by prey identity irrespective of prey size (NODF analysis) but nestedness (in a different sense) could also occur by prey size irrespective of prey identity, right? The same analysis you ran for Q1 with prey mass SD as the response variable could tell you this perhaps. In other words, larger H. venatoria spiders may have a wider prey mass SD bc they like to eat potato chips (small prey) and steaks (large prey; though they tend towards steaks - results from Q1), while smaller H. venatoria spiders still eat potato chips (small prey), but eat a different brand of potato chips than larger H. venatoria, hence, you would not see an effect of nestedness by prey identity with a NODF analysis but could see an effect with an LMM. Let's hope that made sense... 
  
  
Scatterplot by size

Boxplot by species

Generalists vs. specialists – how big you are vs. who you are as a species – at the species level (avg. size of the species and the species identity). 

Dan things the hierarchical clustering is super interesting

Indices of individual diet specialization (look on Google Scholar)

All the data in one analysis – on prey identity. Unless there are ways to break it down in subsets to test a diff hypothesis (e.g. whether sampling location had a reason).





