---
title: "Analyses_October30"
author: "Ana Miller-ter Kuile"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 2
    df_print: paged
    theme: readable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r, include = FALSE}
package.list <- c("here", "tidyverse", 
                  "glmmTMB", "emmeans",
                  "MuMIn", "DHARMa",
                  "effects", "ggeffects",
                  "patchwork", "calecopal",
                  "gt", "vegan", "remotes",
                  "eulerr", "dummies",
                  "emmeans")

## Installing them if they aren't already on the computer
new.packages <- package.list[!(package.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

## And loading them
for(i in package.list){library(i, character.only = T)}
```


# Inter- and intraspecific body size patterns in individual predator-prey interactions from DNA metabarcoding

The analyses for these projects have been inspired by ideas and approaches outlined in: [Brose et al. 2019](https://doi.org/10.1038/s41559-019-0899-x), [Rudolf et al. 2014](https://doi.org/10.1098/rspb.2013.3203), and [Woodward and Hildrew 2002](https://doi.org/10.1046/j.1365-2656.2002.00669.x), among others. Specifically, these analyses explore the importance of body size (both within and across species) and species identity in shaping food web patterns. 

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "geophilomorpha.png"))
```

## Introduction

Food webs are regularly built using "nodes" based on size, predator identity, or a combination of size and identity (i.e. "stage structure"). For food webs, nodes are created by a combination of predator species and size and the interactions (links) between predators and their prey items are determined using size-based rules related to gape limitation. Across species, sizes, and environments, there seem to be consistent patterns of size relationships in food webs - for example, variation in the relationships between predator and prey sizes can be explained by predator traits (e.g. locomotion, thermoregulation) across 290 food webs (Brose et al. 2019). While these patterns are promising for the efficient construction of food webs across many orders of magnitude of body size, they have still not been validated by real data for many species of smaller body sizes, including invertebrates in terrestrial environments. 

In several aquatic field and mesocosm studies, it has been clear that both the body size and species identity of predators is important in determining both the identity and body size of prey items (e.g. Woodward and Hildrew 2002, Rudolf et al. 2014). However, for terrestrial predators, which comprise (%) of total species on the planet, determining these size rules of predation based on observed interactions has been challenging or impossible. These consumers, thus, constitute a data gap of predator-prey interactions built on observed interaction data.

To aim to fill this gap in observed interactions for invertebrate terrestrial predators, in this study, we employed DNA metabarcoding of the COI gene region using general primers for all animals (following an adapted protocol from Krehenwinkel et al. 2016 and Miller-ter Kuile et al. *in review*) to determine the diets of nine species of invertebrate predators on Palmyra Atoll, Central Tropical Pacific. We determined interactions for 182 predator individuals, encompassing 3.5 orders of magnitude of body size and including 339 distinct interactions. Using individual-level body size data for each predator individual and averaged body size data for prey items (similar to resolution for prey in other food webs e.g. Brose et al. 2019), we asked a set of questions related to body size within and across species in food webs. Specifically,

## The questions

1. Interspecific 1: Does predator size, species identity, or their combination determine prey size?

2. Interspecific 2: Does predator size or identity determine prey identity?

3. Intraspecific 1: Do the range of prey sizes vary by individual predator size within a species?

4. Intraspecific 2: Do smaller individuals eat a subset of the prey richness of larger predator individuals (e.g. nestedness or a "nested hierarchy", Woodward and Hildrew 2002) 

5. Intraspecific 3: Is there evidence of stage structure within predator species based on individual body size?

These questions are key to building multi-species food web models that incorporate realistic body size and species dependent feeding habits both across and within predator species.

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "neoscona.png"))
```

## The data

This dataset represents interactions between individual predators and their prey. The predators range in size from $2.3 x 10^{-1}$ to $9.3 x 10^2$ (3 orders of magnitude). Predators represent 9 species, including five species of spider (*Heteropoda venatoria*, *Neoscona theisi*, *Ooonopidae sp.*, *Scytodes longipes*, *Smeringopus pallidus*), one centipede (*Geophilomoropha sp.*), one earwig (*Euborellia annulipes*), a predatory katydid (*Phisis holdhausi*), and a dragonfly species (*Pantala flavescens*).

```{r, include = FALSE}
source(here("6_statistics", "1_pred_prey_size.R"))
```

### Predator size distribution

```{r}
pred_size2
```

### Predator size distribution by species

```{r}
pred_labels <- c("CEN" = "Geophilomorpha sp.", "EUB" = "E. annulipes", 
                 "HEV" = "H. venatoria", "LRS" = "Oonopidae sp.", 
                 "NEO" = "N. theisi", "PAN" = "P. flavescens",
                 "PHH" = "P. holdhausi", "SCY" = "S. longipes",
                 "SME" = "S. pallidus")

size %>%
  distinct(sample, sample_str, pred_mass_mg) %>%
  mutate(sample_str = fct_reorder(sample_str, pred_mass_mg, .fun='mean')) %>%
  ggplot(aes(x = pred_mass_mg, fill = sample_str)) +
  geom_histogram(color = "black", alpha = 0.85) +
  theme_bw() + 
  scale_x_log10() +
  labs(x = "Predator mass (mg)", y = "Individuals") +
    theme(axis.text = element_text(size =20),
          axis.title = element_text(size = 25),
          axis.text.x = element_text(angle = 45, hjust =1)) +
  facet_wrap(~sample_str, labeller = labeller(sample_str = pred_labels)) +
  scale_fill_manual(values = pal_kelp,
                    labels = pred_labels)
```

I compiled prey DNA data at the family level because of resolution in online taxonomy databases. Averaging prey identity and size at the family level is common for food webs of interactions with terrestrial invertebrates (e.g. Brose et al. 2019). The prey family range in size from $6.3 x 10^{-4}$ to $3.1 x 10^2$ (6 orders of magnitude). 
 
### Prey size distribution

```{r}
prey_mean_size2
```

Prey comprise 57 different families of invertebrate organisms. 

```{r}
size %>%
  distinct(Class, Order, Family) %>%
  arrange(Class, Order, Family)
```

## Interspecific patterns

Most food web body size studies focus on large, interspecific relationships with body size and interactions (e.g. Brose et al. 2019, OTHERS). Since food webs are often built based on either species-level nodes, size-based nodes, or some combination of these, understanding how species and size may alter these relationships is key to scaling up food web body size patterns to build realistic depictions of interactions in multi-species communities. In the next two questions, I explore the relationships between predator size and predator species and prey size and identity to try to understand whether any simple "rules" exist for these feeding interactions.

## Q1: Interspecific predator-prey size

### Does predator size, species identity, or their combination determine prey size?

For this question, I used LMMs with log-transformed size values (due to data distribution) and performed model selection using AIC comparing models to a full model that included the interaction between predator body size and species identity (prey_size ~ predator_size x predator_species).

The interpretation of each model structure is as follows:

1. Interaction model (prey_size ~ predator_size x predator_species): Both predator size and species identity determine prey size and both the **slope** and **intercept** of this model vary by predator species.

2. Size + species model (prey_size ~ predator_size + predator_species): Both predator size and species identity determine prey size and only the **intercept** of this model varies by predator species.

3. Size model (prey_size ~ predator_size): There is a relationship between prey size and predator size and *predator species does not change this relationship*.

4. Species model (prey_size ~ predator_species): *Regardless of predator size*, each predator species eats a distinct size range of prey. 

The best model based on AIC was the Size + Species model, which means that there is a significant (positive) relationship between prey size and predator size; the slope of this line does not change across predator species, but each species has a different intercept for this relationship. This means that larger individuals eat larger prey (with a log-log relationship in this case of $y = x^{0.41}$), while individuals within some species have proportionally larger or smaller prey sizes in relation to their body size.

### Size relationship

```{r}
size_graph_col
```

### Species relationship

```{r}
species_graph
```

### Species pairwise differences

```{r}
pairwise_sp
```

#### Predator-prey size interpretations

Both predator species and predator size is an important determinant of prey size, with an invariant slope of this relationship with varying intercepts by species. This means that when building food webs, both species and body size are important determinants of deciding on nodes and links at the species or species-stage level. The species-level differences suggest that some predator species eat relatively smaller or larger prey items than would be expected. In this case, *Euborellia annulipes* (Order: Dermaptera) eats relatively large prey items, which could be because this species uses its large cerci (forcep-like pincers) to handle relatively large prey items or could be evidence of scavenging already-dead prey items in this species. Also, both *Geophilomorpha sp.* (centipede) and *Pantala flavescens* (Order: Odonata) have relatively smaller prey items proportional to other species, suggesting these two species are more gape-limited than the other predators in this study. These patterns suggest a neat "next step" for this kind of work inspired by e.g. Schmitz 2009 relating to predator foraging mode and their ecological impacts, in this case in terms of their rates of biomass consumption.


```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "goblin.png"))
```

## Q2: Interspecific prey composition

### Does predator size or identity determine prey identity?

Next, I wanted to know whether prey item identity is determined by predator size or predator species identity. For this analysis, since predators were sampled from different habitats on the island (e.g. soil and canopy, different forest types) I also included sampling environment as a covariate in this analysis. 

```{r, include = FALSE}
source(here("6_statistics", "2_cca_prey_identity.R"))
```


To determine whether these predators were partitioning prey by size, species, or both, I ran a constrained canonical analysis (CCA) with both predator species identity and size as predictor variables, along with habitat (e.g. forest type) and microhabitat (e.g. canopy vs. understory) as covariates. CCA is a multivariate method that tests for relationships between predictor variables and a set of multivariate data (e.g. a set of diet "communities" across individual predator samples). Unlike RDA, CCA does not assume a linear relationship and is more aimed at understanding how variables influence community composition. 

```{r}
#ANOVA of whole model
anova(cca, permutations=10000)

#how much variation explained in this RDA
RsquareAdj(cca)
```

When examining the terms of this model, all variables are important for explaining variation in these data.

```{r}
#ANOVA of model terms
anova(cca, by='margin', permutations=10000)
```

```{r, eval = FALSE}
summary(cca)
```


From this output, we see that this CCA explains 22.97% of the variation in the data. Of this variation, CCA1 explains 14.4% of the variation and CCA2 explains 12.9% of this variation. 

### CCA and of prey composition by predator

```{r}
CCA_plot
```

We can go further and visualize the variation due to each predictor using an Euler plot, which is like a Venn diagram in that it shows the contribution to variation in each variable independent of and dependent on other variables. 

```{r}
euler
```

The Euler plot shows that predator species explains 12% of the variation alone, habitat explains 4% alone, and microhabitat and predator mass contribute another 1% each. There is some overlap with all these less important variables with predator species, explaining some of the variation in combination with predator species. In total, predictors explain ~25% of the variation in the data.

#### Predator-prey identity and size interpretation

Predator species is the most important predictor of prey species composition/identity. Habitat (e.g. forest type) and microhabitat (e.g. soil vs. canopy) explained some variation, though predator size seems less important in determining prey species composition. The result of this CCA indicates that  predators seem to be partitioning resources by species, regardless of size or habitat use.

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "phisis.png"))
```

## Intraspecific patterns

Up to this point, I have been approaching questions across species to understand the role of body size and species identity in food web structure. Populations of predator species also have feeding interactions that are governed by body size variation within a given species (e.g. Woodward and Hilldrew 2002 and others). Body size variation within predator species is key to determining the level at which nodes should be assigned in a food web and may demonstrate diet partitioning within species based on prey size or prey identity. In many consumers, there is an apparent nested hierarchy or nestedness to intraspecific diet in which smaller individuals eat a subset of the same prey items that larger individuals of that species consume. This is one of the theoretical bases of the niche model for building null food webs (CITE these).

I will use data from the three species with the largest sample sizes within our dataset for the following analyses. These species are *H. venatoria*, *N. theisi*, and *P. holdhausi*. Similar to intraspecific diet variation, I will be examining both the size and composition of prey items in these predators to determine whether these species partition prey items in a predictable way throughout their lifespans.

## Q3: Intraspecific range of prey size by predator size

### Do the range of prey sizes vary by individual predator size within a species

For this first **prey size range** I used linear models to examine the prey items within individual predators in a species to determine whether larger individuals eat a wider range of prey sizes than smaller individuals. For this analysis, I was limited to individuals that had consumed at least 2 prey items and used the range of size of prey as a response variable and predator mass as a predictor variable. I log transformed both the response and predictor variables due to data distributions and ran all species in one model with an interaction between predator size and species (log_prey_size_range ~ log_predator_size*predator_species) to avoid issues with multiple comparisons. 

```{r, include = FALSE}
source(here("6_statistics", "3_prey_size_range.R"))
```

The resulting model indicated a significant relationship between predator size and the range of diet items that predator consumed. Interestingly, this relationship was almost linear and positive for the species with the largest size range (*H. ventatoria*, $y = x^{0.99}), whereas the relationship demonstrated exponential decay for the species with smaller size ranges. Importantly, these two relationships seem to be driven by 1-2 samples in both species and would otherwise suggest very little or no relationship between predator size and prey size range.

#### Prey size range by predator size

```{r}
range_graph
```

#### Interpretation of prey size range by predator size

For one species (*H. venatoria*), it does seem that larger predators will eat a broader range of prey sizes. However, this relationship does not hold for two other species (*N. theisi* and *P. holdhausi*). *H.venatoria* has the largest range in minimum to maximum body size (`r min(HEV$pred_mass_mg)` -`r max(HEV$pred_mass_mg)`mg) of the three species, which may indicate that whether larger individuals eat larger prey than smaller individuals is dependent on the difference in size between those subsets of the predator population. 

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "euborellia.png"))
```

## Q4: Intraspecific nestedness of prey identity

### Within a species, do smaller individuals eat a subset of the prey of larger individuals (e.g. nestedness)?

In addition to predators within a population partitioning resources based on resource (prey) size, they may also partition resources based on prey identity. To explore this, I did two related analyses. The first was to determine whether individual diets within these predator individuals illuminate nestedness or a nested hierarchy. Phrased differently, nestedness determines whether the diet of some individuals in a population (smaller individuals) is a subset of the diet of other (larger) individuals. To do this nestedness analysis, I asked whether the diets of individuals were nested compared to random simulations of the same community (I used the NODF nestedness measure, Ulrich et al. 2009), which incorporates pairwise similarities across sites and species occurrences and is invariant to the size of the sample or community sampled. Specifically, the NODF nestedness value can take community assemblies and ask whether the most species poor of these comprise a subset of the species that are present in a richer community (Almeida-Neto et al. 2008). 

```{r, include = FALSE}
source(here("6_statistics", "4_intraspecific_nestedness.R"))
```

### *H. venatoria* nestedness analysis

```{r}
oecosimu(mat_hev, nestednodf, "quasiswap")
```

The results of this simulation indicate that the prey communities across *H. venatoria* individuals are not significantly nested compared to randomly-constructed communities.

A visualization of this:

```{r}
hev_matrix 
```

And suggestion that we sampled an estimated 28 of an estimated 42 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
hev_acc
```

### *N. theisi* nestedness analysis

```{r}
oecosimu(mat_neo, nestednodf, "quasiswap")
```

Again, the results of this simulation indicate that the prey communities across *N. theisi* individuals are not significantly nested compared to randomly-constructed communities.

```{r}
neo_matrix
```

And suggestion that we sampled an estimated 29 of an estimated 44 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
neo_acc
```

### *P. holdhausi* nestedness analysis

```{r}
oecosimu(mat_phh, nestednodf, "quasiswap")
```

Again, the results of this simulation indicate that the prey communities across *P. holdhausi* individuals are not significantly nested compared to randomly-constructed communities.

```{r}
phh_matrix
```

And suggestion that we sampled an estimated 18 of an estimated 35 prey species in the prey community based on frequency across samples (Though, I wonder if it makes sense to re-do these analyses with read abundances, Austen?)

```{r}
phh_acc
```

#### Predator-prey identity and size interpretation

None of these predator populations seem to have significantly nested diets within a species. Unlike results from e.g. Woodward and Hildrew 2002, it does not seem like a good assumption on building feeding interactions within species stages would be to assign a subset of prey to smaller individuals than larger individuals. We captured anywhere from 51 - 67% of total prey richness based on accumulation curves and I'm wondering if this is coming into play here. 


```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "pantala.png"))
```

## Q5: Intraspecific stage structure based on size

```{r, include = FALSE}
source(here("6_statistics", "5_stage_structure_cons.R"))
```

### Is there evidence of stage structure within predator species based on individual body size?

Finally, and related to Q4 above, I explored whether, even without nestedness in its pattern, whether there was evidence of stage structure (e.g. niche differentiation) based on body size within these predator populations. I approached this by performing a hierarchical clustering based on Jaccard similarity and using the Unweighted Pair-Group Method Using Arithmetic Means (UPGMA) algorithm, which clusters the most similar communities based on Jaccard similarity and then uses averaged between group differences to create a dendrogram (clustering tree) of increasingly less similar pairs of sites. It is commonly used in ecology to compare the species composition across sites (e.g. Pavao et al. 2019).

I performed this analysis two ways: 1) based on 100% similarity and 2) based on 50% similarity. I performed the analyses both ways because the 100% similarity ended up clustering primarily individuals with diet richness of one species; whereas the 50% clustering was less biased toward low-richness samples. The results were a list of clusters at each similarity cut-off with which I performed GLMs with predator mass as a dependent variable and cluster identity as a predictor variable. 

### *H. venatoria* 100% similarity clusters

Twenty-nine of fifty-three (29/53, 55%) of *H. venatoria* individuals clustered with at least one other individual with 100% similarity. Based on model selection, a model with cluster ID was a better-fitting model than one without cluster ID ($\Delta$AIC = 21.75), however, post-hoc Tukey tests indicated no significant pairwise differences.

```{r}
(dendro_hev) / 
  (clusters_HEV + rich_HEV + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 100% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).

### *N. theisi* 100% similarity clusters

Five of twenty-four (5/24, 21%) of *N. theisi* individuals clustered with at least one other individual with 100% similarity. Predator size did not significantly differ between clusters. 

```{r}
(dendro_NEO) / 
  (clusters_NEO + rich_NEO + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 100% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).

### *P. holdhausi* 100% similarity clusters

Twenty-five of forty-two (25/42, 60%) of *P. holdhausi* individuals clustered with at least one other individual with 100% similarity. Predator size did not significantly differ between clusters. 

```{r}
(dendro_PHH) / 
  (clusters_PHH + rich_PHH + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 100% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).


```{r, include = FALSE}
source(here("6_statistics", "6_stage_structure_big.R"))
```

### *H. venatoria* 50% similarity clusters

Forty-four of fifty-three (44/53, 83%) of *H. venatoria* individuals clustered with at least one other individual with 50% similarity. Predator size did not significantly differ between clusters. 

```{r}
(dendro_hev) / 
  (clusters_HEV + rich_HEV + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 50% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).

### *N. theisi* 50% similarity clusters

Fourteen of twenty-four (14/24, 58%) of *N. theisi* individuals clustered with at least one other individual with 50% similarity. Predator size did not significantly differ between clusters. 

```{r}
(dendro_NEO) / 
  (clusters_NEO + rich_NEO + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 50% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).

### *P. holdhausi* 50% similarity clusters

Thirty-eight of forty-two (38/42, 90%) of *P. holdhausi* individuals clustered with at least one other individual with 50% similarity. Predator size did not significantly differ between clusters.

```{r}
(dendro_PHH) / 
  (clusters_PHH + rich_PHH + plot_layout(widths = c(3,1))) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 25))
```

A dendrogram (A) with dotted line indicating clusters with 50% similarity. Predator mass within each cluster (B), and the number of prey items in each of these clustered predators (C).

#### Stage structure interpretation

There is no evidence of clear stage structuring by body size in any of these predator species. Thus, it does not appear that these species display clear, discrete "stages" like those of species that may go through complete metamorphosis. In this case, I think it is important to note the sparsity of interactions in our interaction matrix, with many predator individuals only having a few prey items in their diet during collection, suggesting that this may be an analysis worth revisiting with a larger sample size within predators with these sparse diet matrices. 

```{r, out.width="20%", fig.align="right"}
knitr::include_graphics(here("art", "png", "heteropoda.png"))
```

## Final summary

#### Interspecific variation

**Interspecific predator-prey body size relationships** follow a predictable relationship of prey_size = predator_size$^{0.41}$. This relationship is mediated by predator species, suggesting one or more life history traits that may mediate this relationship, including, in this case, the potential for tools or scavenging to increase prey size relative to predator size (*E. annulipes*, Dermaptera) or by decreasing prey size relative to predator size due to stronger gape limitation (*P. flavescens* and Geophilomorpha sp.). Future work in body size relationships could focus on determining these trait-mediated effects (e.g. Brose et al. 2019).

**Interspecific prey identity** is determined primarily by predator identity, with some effect of the size of predators and habitat location. This means that predator species seem to partition prey by species, rather than size classes. 

#### Intraspecific variation

**Intraspecific range of prey size** increases for larger individuals of one predator species (*H. venatoria*), and not for two other species (*N. theisi* and *P. holdhausi*). *H.venatoria* has the largest range in minimum to maximum body size of the three species, which may indicate that whether larger individuals eat larger prey than smaller individuals is dependent on the difference in size between those subsets of the predator population. 

**Intraspecific nestedness of prey identity** is not evident for the three predators in this analysis (*H. venatoria*, *N. theisi*, and *P. holdhausi*). Unlike results from e.g. Woodward and Hildrew 2002, it does not seem like a good assumption on building feeding interactions within species stages would be to assign a subset of prey to smaller individuals than larger individuals. We captured anywhere from 51 - 67% of total prey richness based on accumulation curves and I'm wondering if this is coming into play here.

**Intraspecific stage structure based on size** does not seem to be evident in these predator species. Thus, it does not appear that these species display clear, discrete "stages" like those of species that may go through complete metamorphosis. In this case, I think it is important to note the sparsity of interactions in our interaction matrix, with many predator individuals only having a few prey items in their diet during collection, suggesting that this may be an analysis worth revisiting with a larger sample size within predators with these sparse diet matrices. 

#### Next steps/caveats?




